man/ts-warp.5
.\" Copyright (c) 2021-2026, Mikhail Zakharov <zmey20000@yahoo.com>
.\"
.\" Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
.\" following conditions are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
.\"    the following disclaimer in the documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH TS\-WARP 5 "June 6, 2022" "1.0" "Mikhail Zakharov"
.SH NAME
\fBTS\-Warp\fR INI-like configuration file format
.SH DESCRIPTION
\fBts\-warp\fR uses INI-like configuration files to identify SOCKS servers for destinations that clients are willing to
reach. The format of ts-warp INI-files are almost dentical to the classical INI dialect, with the only main exception,
that duplicate names for \fBkeys\fR are alloved within the same \fBsection\fR.

Each \fBsection\fR, a string, enclosed in square brackets ('[', ']'), specifies a name for a SOCKS server block. A
section consists of keys (variables) describing \fBSOCKS server\fR attributes, as well, as keys specifying destination
\fBtargets\fR.

Remarks are specified with: \fB;\fR and \fB#\fR characters

Key names are case insensitive i.e., \fBtarget_host\fR is the same as \fBtArGeT_HoSt\fR
.SH \fBSOCKS server\fR keys
.TP
.B socks_server = <IP | Hostname>[:port]
Manadatory variable, specifying a SOCKS server for the section. Optional port number defaults to 1080.
.BR
.TP
.B socks_chain = SECTION_1, SECTION_N
Defines a chain of sections with SOCKS servers, that are required to reach first before contacting the server in the
current section. Optional variable.
.BR
.TP
.B socks_version = 5 | 4
Specifies the server SOCKS version. Optional, defaults to 5.
.BR
.TP
.B socks_user = username
A username if the server requires authentication. Only plaintext authentication is supported.
.BR
.TP
.B socks_password = plain:PlaintextPassword | tsw01:HEXADECIMALHASH
A user password if the SOCKS-server requires authentication. Can be specified as plaintext if preceded with \fBplain:\fR
prefix or encoded by \fBts-pass(8)\fR with TS-Warp obfuscation algorithm if the \fBtsw01:\fR prefix is set.
.SH \fBTarget\fR keys
.TP
.B target_host = <IP | Hostname>:[port_1[-port_n]]
Describes a single target host either with it's IP address or resolveable host name. Optionally a port or a port range
can be specified to narrow the band.
.BR
.TP
.B target_domain = example.org:[port_1[-port_n]]
Specifies the name of a domain resolveable via DNS. Optionally a port or a port range can be specified to narrow the
band.
.BR
.TP
.B target_network = IP[:port_1[-port_n]]/Netmask
Specifies an IPv4 or IPv6 subnet. Netmasks are accepted only in dot-decimal notation. Optionally a port or a port range
can be specified to narrow the band.
.BR
.TP
.B target_range = IP_1[:port_1[-port_n]]/IP_n
Specifies a range of sequential IPv4, IPv6 addresses. Optionally a port or a port range can be specified to narrow the
band.
.SH EXAMPLES
.EX
[HOME NETWORK]                          ; Section definition
# Target definitions:
; target_host = 192.168.1.1
; target_host = Anvil
; target_network = 192.168.1.0/255.255.255.0
; target_network = 192.168.1.0/24
; target_range = 192.168.1.1/192.168.1.20
target_domain = balmora.lan

# Server definitions:
socks_server = 192.168.1.237
; socks_version = 5
socks_user = socks
; socks_password = plain:TopSecretPass@34
socks_password = tsw01:08415D5F6519633F1D150E08552837506D12383C177C176F7C322E1F562D
.EE
.SH SEE ALSO
.BR ts\-warp (8),
.BR ts\-warp.sh (1),
.BR ts\-pass (1),
.BR gui\-warp (1)


CHANGELOG.md
# CHANGELOG

* **Current**
  * `ts-warp`: Introduced `section_balance = disabled` to exclude sections from target to section matching

* **2026.02.15    ts-warp-1.5.9, gui-warp-1.0.28 (gui-warp-v1.0.35-mac), ns-warp-1.0.7**
  * `gui-warp` and `gui-warp.app`: ACT memory optimizations
  * `ts-warp.c`: `gcc` `warn_unused_result` fixed for `seteuid(pwd->pw_uid)`
  * `gui-warp.py`: memory leak fix - some objects were never released due to refresh/timers issues
  * `gui-warp.app`: `build.sh` updated to save `iSSH2` logs on build-fail
  * `gui-warp.app`: memory leak fix - some objects were never released due to refresh/timers issues
  * `gui-warp.app`: `ts-warp.sh act` option to trigger connection statistics dump
  * `gui-warp.app`: `build.sh` clean after build
  * `gui-warp.app`: `iSSH2-head.sh` updated to version `1.0`
  * `gui-warp.app`: `build.sh` enhanced; `iSSH2-head.sh` included
  * `ts-warp.c`: Typo errors fixed
  * `git`: Performed `git-filter-repo` to clean garbage.
  * `configure`: Use `xcrun` to detect MacOS SDK path. Fixes the [Issue](https://github.com/mezantrop/ts-warp/issues/14)
  * `gui-warp.app`: More friendly `ts-warp.sh` for CLI execution on `MacOS`
  * `ts-warp.c, pidlist.c`: better `PID`-list management
  * `inifile.c`: suppress negative name lookup messages
  * `gui-warp.py`: INI-tab Text-field resizing corrected

* **2025.06.04    ts-warp-1.5.8, gui-warp-1.0.27 (gui-warp-v1.0.34-mac), ns-warp-1.0.7**
  * `build.sh`: `libssl/libssh` section tweaks
  * `ts-warp.c`: `SIGHUP` hooked to recreate/reopen the log-file
  * `ts-warp.c`: trap signals for the `console` mode as well
  * `minmiconf:` Updated to version `1.1.1.1`
  * `Makefile`: updated to fail `make install` if `./configure` and `make all` were not run
  * `configure`: Do not fail on macOS SDK absence, if `gcc` is detected on `macOS`
  * `inifile.c`: "Long" dotted-decimal netmasks style returned, so `255.255.0.0` and `/16` both working
  * `inifile.c`: ignore netmasks for `target_host`
  * `inifile.c`: correct `/32` netmask for `target_network`
  * `gui-warp.py`/`gui-warp.py.in`: 'Password hash' caption
  * `CONTRIBUTORS.md`: New contributor Marcin Nowicki <marcin.nowicki@kyndryl.com>. Thanks for your ideas and testing!
  * `build.sh`: `-no-tests` added for OpenSSL `Configure`; `-fembed-bitcode` removed for both OpenSSL and libssh2
  * `gui-warp.py`: Highlight "Save" button for INI/FW if config has been edited. Thanks Radek Sipka for the request

* **2024.11.08    ts-warp-1.5.7, gui-warp-1.0.26 (gui-warp-v1.0.33-mac), ns-warp-1.0.7**
  * `ts-warp.sh`: Log-file truncating moved here
  * `gui-warp.py`: Minor changes about log-file size and FW-file selection under `Linux`
  * `ts-warp.c`: Null-pointer dereference in `PIDS` table processing fixed
  * `ts-warp.c`: `ACT` reporting `[SECTIONS]`
  * `inifile.c`: Correct usage of multiple links of a proxy-chain
  * `inifile.c`: Potential buffer-overflow fixed in parsing `[SECTIONS]`
  * `ts-warp.c`: Proxy server selection optimization
  * `pidlist.c`: `Uninitialised value was created by a heap allocation` fixed

* **2024.09.07    ts-warp-1.5.6, gui-warp-1.0.25 (gui-warp-v1.0.32-mac), ns-warp-1.0.7**
  * `GUI-Warp.app`: Better handling of `starter`, `starter.c` updated to v1.3
  * `GUI-Warp.app`: `starter.c` updated to v1.2
  * `inifile.c`: `proxy_server` variable parsing. Thanks <jesussanchezvalle@gmail.com> for the bug-hunting and testing
  * `GUI-Warp.app`: `starter.c` to handle environment variables
  * `minmiconf`: Updated to version `1.1.1.0`
  * `configure`: Added lost paths check for macOS SDK directories
  * `GUI-Warp.app`: Added an internal launcher - `starter.c`
  * `ts-warp.sh`: Enable `PF`-firewall in `ts-warp.sh` `start`; Minor fixes for `PF` on `macOS` in `GUI-Warp` on `stop`
  * `pidlist.c`: `ACT` shows destination IP-address, when internal proxy servers are used
  * `ts-warp.c`: `-D 0..512` DPI bypass fragment size by default 0 - disabled; To enable use a positive value, e.g., 2

* **2024.08.15    ts-warp-1.5.5, gui-warp-1.0.25 (gui-warp-v1.0.30-mac), ns-warp-1.0.7**
  * `ts-warp.c`, `http.c`: Deep Packet Inspections bypass. Option `-D` to disable it.
  * `ts-warp.c`: Internal proxy servers allowed making direct connection to destinations

* **2024.07.23    ts-warp-1.5.4, gui-warp-1.0.25 (gui-warp-v1.0.30-mac), ns-warp-1.0.7**
  * `ts-warp.c`: `ACT`-file created as `RUNAS_USER` user owner
  * `ts-warp.c`: On `macOS` delayed `setuid()`/`setgid()` disabled as almost useless

* **2024.07.23    ts-warp-1.5.3, gui-warp-1.0.25 (gui-warp-v1.0.30-mac), ns-warp-1.0.7**
  * `ts-warp.c`: Enable on `macOS` delayed `setuid()`/`setgid()` to run as non-privileged user
  * `ts-warp.c`: SSH2 proxy connection fixed, thanks Stefan Hildebrandt for the bug-hunting
  * `ssh2.c`: `password` for `keyboard-interactive` and `password` now always required
  * `examples\ts-warp.ini`: Highlight info about accessing `proxy_key` (the private key) permissions/location

* **2024.07.12    ts-warp-1.5.2, gui-warp-1.0.25 (gui-warp-v1.0.30-mac), ns-warp-1.0.7**
  * `GUI-Warp`: `About`-tab displays `CHANGELOG.md`-file
  * `Makefile`: installs/deinstalls text-files files into/from `<PREFIX>/share/ts-warp` directory
  * `GUI-Warp.app`: `ts-warp.sh` can be run from `CLI` to control `ts-warp` daemon e.g.:
      `sudo /Applications/gui-warp.app/Contents/Resources/ts-warp.sh start /Users/$USER/ts-warp`
  * `GUI-Warp.app`: `build.sh` uses current `openssl` and `libssh2` versions
  * `Makefile`: `WITH_LIBSSH2?=0` to disable `LIBSSH2` by default
  * `ts-warp_autofw.sh` Ignore domain names if they are not resolveable to IP-addresses
  * Typos fixed again
  * `gui-warp.py`: Minor cosmetic changes
  * `minmiconf`: `v1.1.0.2` - OpenBSD support; `ts-warp_auto.sh` modernization; `gui-warp.py`: Save INI button fix
  * `gui-warp.py`: `v1.0.30-mac`, `gui-warp.py.in`: `v1.0.25`, `minmiconf`: `v1.1.0.1`
  * `configure`: updated to `minmiconf-1.1.0.0`
  * Ignore a section if `proxy_server` value is not specified


gui/ts-warp.py
#!/usr/bin/env python3

"""
--------------------------------------------------------------------------------
TS-Warp - Transparent SOCKS protocol Wrapper GUI frontend
--------------------------------------------------------------------------------

Copyright (c) 2021, 2022, Mikhail Zakharov <zmey20000@yahoo.com>

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import tkinter as tk
import tkinter.ttk as ttk
import subprocess


class App:
    def __init__(self, root, startcmd, 
                 logfile='/usr/local/var/log/ts-warp.log',
                 pidfile='/usr/local/var/run/ts-warp.pid'):

        root.title("TS-WARP GUI v0.1")
        width = 800
        height = 560
#        screenwidth = root.winfo_screenwidth()
#        screenheight = root.winfo_screenheight()
        root.geometry(f'{width}x{height}')
        root.resizable(width=True, height=True)

        root.columnconfigure(0, weight=1)
        root.rowconfigure(0, weight=0)
        root.rowconfigure(1, weight=1)
        root.rowconfigure(2, weight=0)

        # Top button frame
        lfrm_top = tk.LabelFrame(root, height=40, relief=tk.FLAT, padx=2, pady=4)
        lfrm_top.grid(column=0, row=0, sticky=tk.N + tk.W + tk.E)
        lfrm_top.columnconfigure(0, weight=0)
        lfrm_top.columnconfigure(1, weight=0)
        lfrm_top.columnconfigure(2, weight=0)

        btn_run = tk.Button(lfrm_top, width=2, height=1, text='▶')
        btn_run.grid(row=0, column=0, sticky=tk.W, padx=2)
        # btn_run['command'] = self.start

        btn_rld = tk.Button(lfrm_top, width=2, height=1, text='⟳')
        btn_rld.grid(row=0, column=1, sticky=tk.W, padx=2)

        # Display config/log pane
        tabControl = ttk.Notebook(root)
        tab_ini = ttk.Frame(tabControl)
        tab_log = ttk.Frame(tabControl)

        tabControl.add(tab_ini, text='Configuration')
        tabControl.add(tab_log, text='Log')
        tabControl.grid(column=0, row=1, sticky=tk.NSEW)

        # Tab INI/Config
        tab_ini.columnconfigure(0, weight=1)
        tab_ini.rowconfigure(0, weight=1)

        ini_txt = tk.Text(tab_ini)
        ini_txt.grid(column=0, row=0, sticky=tk.NSEW)

        scroll_ini = tk.Scrollbar(tab_ini, orient=tk.VERTICAL)
        scroll_ini.grid(column=1, row=0, sticky=tk.NSEW)
        scroll_ini.config(command=ini_txt.yview)
        ini_txt.config(yscrollcommand=scroll_ini.set)

        # Tab Log
        tab_log.columnconfigure(0, weight=1)
        tab_log.rowconfigure(0, weight=1)

        log_txt = tk.Text(tab_log)
        log_txt.grid(column=0, row=0, sticky=tk.NSEW)
        log_txt.config(state='disabled')
        tab_log.bind("<Visibility>", self.readfile(log_txt, logfile))

        scroll_log = tk.Scrollbar(tab_log, orient=tk.VERTICAL)
        scroll_log.grid(column=1, row=0, sticky=tk.NSEW)
        scroll_log.config(command=log_txt.yview)
        log_txt.config(yscrollcommand=scroll_log.set)

        # Status bar
        lfrm_bottom = tk.LabelFrame(root, relief=tk.FLAT, padx=2)
        lfrm_bottom.grid(column=0, row=2, sticky=tk.S + tk.W + tk.E)
        lfrm_bottom.columnconfigure(0, weight=1)

        lbl_stat = tk.Label(lfrm_bottom, text='■ running', fg='green')
        lbl_stat.grid(row=0, column=0, sticky=tk.E)
        self.status(lbl_stat, pidfile)

    def readfile(self, t_widget, logfile):
        t_widget.config(state='normal')
        lf = open(logfile, 'r')
        t_widget.delete(1.0, tk.END)
        t_widget.insert(tk.END, lf.read())
        t_widget.see(tk.END)
        lf.close()
        t_widget.config(state='disabled')
        root.after(1000, self.readfile, t_widget, logfile)

    def status(self, t_widget, pidfile):
        pf = None
        try:
            pf = open(pidfile, 'r')
        except:
            t_widget['text'] = '■ Stopped'
            t_widget['fg'] = 'red'

        if pf:
            t_widget['text'] = '■ Running: ' + pf.readline()[:-1]
            t_widget['fg'] = 'green'
            pf.close()

        root.after(1000, self.status, t_widget, pidfile)

#    def start(self):
#        subprocess.run([self.startcmd, 'start'])


if __name__ == "__main__":
    root = tk.Tk()
    app = App(root, startcmd='/usr/local/etc/ts-warp.sh', logfile='/var/log/ts-warp.log', pidfile='/var/run/ts-warp.pid')
    root.mainloop()
